File: src/lib/runManifest.ts
/**
 * runManifest.ts — §12 Reproducibility: run manifest 타입 + 로더
 *
 * 백엔드 `kuma_core/shared/run_manifest.py` 와 스키마를 공유한다.
 * `SCHEMA_VERSION` 상수를 이 파일 한 곳에만 정의하며,
 * 백엔드 schema_version 변경 시 이 상수도 동기화해야 한다.
 *
 * 관련: docs/standards/common-frontend-standards.md §12
 */

import { readTextFile } from "@tauri-apps/plugin-fs";

// ── 스키마 버전 (backend 동기화 단일 포인트) ─────────────────────────────────
export const SCHEMA_VERSION = "1.0";

// ── 입력 파일 메타 ────────────────────────────────────────────────────────────

export interface RunManifestInput {
  /** 절대 경로 */
  path: string;
  /** SHA-256 hex digest */
  sha256: string;
  /** 파일 크기 (bytes) */
  size_bytes: number;
}

// ── manifest 루트 타입 ────────────────────────────────────────────────────────

export interface RunManifest {
  schema_version: string;
  /** 백엔드 RPC method 이름 */
  method: string;
  kuma_version: string;
  kuro_module_version?: string;
  python_version: string;
  /** "linux" | "macos" | "win32" */
  platform: string;
  started_at: string;
  finished_at: string;
  duration_seconds: number;
  inputs: Record<string, RunManifestInput>;
  params: Record<string, unknown>;
  seed: number | null;
  extra?: Record<string, unknown>;
}

// ── type guard ────────────────────────────────────────────────────────────────

/**
 * value 가 RunManifest 구조인지 확인하는 type guard.
 * 필수 최상위 필드만 검사한다 (과도한 재귀 검증 회피).
 */
export function isRunManifest(value: unknown): value is RunManifest {
  if (typeof value !== "object" || value === null) return false;
  const v = value as Record<string, unknown>;
  return (
    typeof v["schema_version"] === "string" &&
    typeof v["method"] === "string" &&
    typeof v["kuma_version"] === "string" &&
    typeof v["python_version"] === "string" &&
    typeof v["platform"] === "string" &&
    typeof v["started_at"] === "string" &&
    typeof v["finished_at"] === "string" &&
    typeof v["duration_seconds"] === "number" &&
    typeof v["inputs"] === "object" &&
    v["inputs"] !== null &&
    typeof v["params"] === "object" &&
    v["params"] !== null &&
    (v["seed"] === null || typeof v["seed"] === "number")
  );
}

// ── 로더 ─────────────────────────────────────────────────────────────────────

/**
 * Tauri plugin-fs 로 경로를 읽고 JSON 파싱 + 스키마 검증을 수행한다.
 *
 * @throws 파일 읽기 실패, JSON 파싱 오류, 스키마 불일치 시 Error
 */
export async function loadManifestFromFile(path: string): Promise<RunManifest> {
  let text: string;
  try {
    text = await readTextFile(path);
  } catch (cause) {
    throw new Error(`manifest 파일을 읽을 수 없습니다: ${path}\n${String(cause)}`);
  }

  let parsed: unknown;
  try {
    parsed = JSON.parse(text) as unknown;
  } catch (cause) {
    throw new Error(`manifest JSON 파싱 실패: ${path}\n${String(cause)}`);
  }

  if (!isRunManifest(parsed)) {
    throw new Error(
      `manifest 구조가 올바르지 않습니다. 필수 필드가 누락되었을 수 있습니다: ${path}`,
    );
  }

  if (parsed.schema_version !== SCHEMA_VERSION) {
    throw new Error(
      `지원하지 않는 manifest schema_version: "${parsed.schema_version}". ` +
        `현재 지원 버전: "${SCHEMA_VERSION}". ` +
        `kuma 를 업데이트하거나, 이전 버전으로 생성된 manifest 인지 확인하세요.`,
    );
  }

  return parsed;
}

File: kuma_core/shared/run_manifest.py
"""Reproducibility manifest generation for kuma pipeline runs.

Each export operation writes a ``{output_basename}.run.json`` file alongside
the exported file. The manifest records:

- kuma/kuro/python/platform version
- input file SHA-256 hashes + sizes
- call parameters (sanitised to JSON-serialisable form)
- ISO-8601 UTC timestamps + duration

Usage::

    from kuma_core.shared.run_manifest import (
        compute_input_sha256,
        build_run_manifest,
        write_run_manifest,
        load_run_manifest,
    )

Schema version history:
    1.0  Initial release (2026-05-07)
"""

from __future__ import annotations

import hashlib
import json
import platform
import sys
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from kuma_core.shared.atomic_write import atomic_write_text
from kuma_core.shared.version import KUMA_VERSION, KURO_MODULE_VERSION

SCHEMA_VERSION = "1.0"

# ---------------------------------------------------------------------------
# Public helpers
# ---------------------------------------------------------------------------


def compute_input_sha256(path: Path) -> str:
    """Return the hex-encoded SHA-256 digest of *path*.

    Reads the file in 1 MiB chunks so large files do not exhaust memory.

    Raises:
        FileNotFoundError: *path* does not exist.
        IsADirectoryError: *path* is a directory.
    """
    h = hashlib.sha256()
    with open(path, "rb") as fh:
        for chunk in iter(lambda: fh.read(1 << 20), b""):
            h.update(chunk)
    return h.hexdigest()


def _serialise_params(params: dict[str, Any]) -> dict[str, Any]:
    """Return a JSON-safe copy of *params*.

    Path objects are converted to strings. Non-serialisable objects fall
    back to their ``str()`` representation (no silent data loss — the string
    is always present).
    """
    return json.loads(json.dumps(params, default=str))


def build_run_manifest(
    *,
    method: str,
    inputs: dict[str, Path],
    params: dict[str, Any],
    started_at: datetime,
    finished_at: datetime,
    seed: int | None = None,
    extra: dict[str, Any] | None = None,
) -> dict[str, Any]:
    """Construct the run-manifest dict (not yet written to disk).

    Args:
        method: RPC method name, e.g. ``"design_sdm_primers"``.
        inputs: Mapping from logical name to file ``Path``. Missing or
            non-existent paths are silently omitted from the manifest
            (spec: "입력 파일 부재 시 해당 키 생략").
        params: Handler call parameters. Serialised via ``json.dumps`` with
            ``default=str``; paths within params are normalised to strings.
        started_at: UTC datetime recorded at handler entry.
        finished_at: UTC datetime recorded after export write completes.
        seed: Optional RNG seed for reproducibility.
        extra: Optional free-form dict for handler-specific metadata.

    Returns:
        dict conforming to schema version 1.0.
    """
    # Validate timestamps have timezone info
    if started_at.tzinfo is None:
        raise ValueError("started_at must be timezone-aware (UTC)")
    if finished_at.tzinfo is None:
        raise ValueError("finished_at must be timezone-aware (UTC)")

    # Build inputs section — silently skip absent files.
    inputs_section: dict[str, Any] = {}
    for key, path in inputs.items():
        if not isinstance(path, Path):
            path = Path(path)
        if not path.exists():
            continue
        try:
            sha = compute_input_sha256(path)
            size = path.stat().st_size
        except (OSError, PermissionError):
            continue
        inputs_section[key] = {
            "path": str(path),
            "sha256": sha,
            "size_bytes": size,
        }

    duration = (finished_at - started_at).total_seconds()

    manifest: dict[str, Any] = {
        "schema_version": SCHEMA_VERSION,
        "method": method,
        "kuma_version": KUMA_VERSION,
        "kuro_module_version": KURO_MODULE_VERSION,
        "python_version": sys.version,
        "platform": platform.system().lower(),
        "started_at": started_at.isoformat(),
        "finished_at": finished_at.isoformat(),
        "duration_seconds": round(duration, 3),
        "inputs": inputs_section,
        "params": _serialise_params(params),
        "seed": seed,
    }

    if extra is not None:
        manifest["extra"] = _serialise_params(extra)

    return manifest


def write_run_manifest(output_path: Path, manifest: dict[str, Any]) -> Path:
    """Write *manifest* as JSON to *output_path*.

    The parent directory must already exist (same contract as other kuma
    export helpers — directory creation is the caller's responsibility).

    Args:
        output_path: Destination path; must end in ``.json``.
        manifest: Dict returned by :func:`build_run_manifest`.

    Returns:
        The resolved absolute path that was written.

    Raises:
        ValueError: *output_path* does not have a ``.json`` extension.
        FileNotFoundError: Parent directory does not exist.
    """
    output_path = Path(output_path).resolve()
    if output_path.suffix.lower() != ".json":
        raise ValueError(
            f"Manifest path must end in .json, got: {output_path.suffix!r}"
        )
    if not output_path.parent.exists():
        raise FileNotFoundError(
            f"Parent directory does not exist: {output_path.parent}"
        )
    # Atomic (temp + os.replace) so an interrupted write never leaves a
    # truncated manifest behind the final path.
    atomic_write_text(
        output_path,
        json.dumps(manifest, ensure_ascii=False, indent=2),
    )
    return output_path


def load_run_manifest(path: Path) -> dict[str, Any]:
    """Load and return the manifest dict from *path*.

    Raises:
        FileNotFoundError: *path* does not exist.
        json.JSONDecodeError: File content is not valid JSON.
    """
    path = Path(path)
    if not path.exists():
        raise FileNotFoundError(f"Manifest file not found: {path}")
    with open(path, encoding="utf-8") as fh:
        return json.load(fh)


__all__ = [
    "SCHEMA_VERSION",
    "compute_input_sha256",
    "build_run_manifest",
    "write_run_manifest",
    "load_run_manifest",
]
